docs/design/sync-manager.md

```diff
@@ -269,13 +273,181 @@ manager, we have the following hard requirements:
+it will find no storage, and even though it doesn't sync passwords, it knows
+passwords should be enabled. It will not put passwords in the declined list, but
+neither will it create the passwords collection. So even though the app supplied
+`{passwords: Enabled}`, it will now get back `{passwords: Unknown}`. This will
+only be a problem is Fenix decides to show "passwords" in its "choose what to
+sync" - so Fenix should probably avoid offering unknown engines (which is fine,
+because that would probably confuse the user anyway!) Regardless, in this
+scenario, some other device that *does* sync passwords will show up and will
+remember that it was enabled, so everything ends up OK.
+
+#### How do new engines become enabled?
+
+As implied above, the state of an engine transitions from "Unknown" to "Allowed"
+when it appears in info/collections and isn't in declined. When the very first
+device which knows about a new engine appears, and assuming this device hasn't
+already shown explicit UI allowing the new engine, we can expect the application
+to specify "Unknown" for the new collection. During the sync, we can expect that
+collection to not appear in info/collections, so the sync manager will return
+"Unknown" for this engine - and it will not sync.
+
+At some point, this app will show UI asking the user if they want to sync
+this new engine (this need not be special UI, the normal "choose what to sync"
+UI for the app will presumably include this new engine). After that explicit
+UI, the app will begin supplying either "Allowed" or "Declined" for this engine,
+and the sync manager will take the appropriate action (ie, either add it to
+declined, or sync it causing info/collections to begin reporting it)
+
+#### Problems, challenges, etc
+
+There's a UX issue here in that applications somehow needs to handle this
+tri-state value. In most cases the app will arrange to only show a
+"choose what to sync" UI after connecting and after the first sync, but
+it might also need to handle "incidental" showing of this UI before that
+first sync completes - for example, simply *showing* the UI probably shouldn't
+force changing the state of an engine from "Unknown". The UX challenge is to
+ensure the user really has consented before marking an engine as "Allowed".
+
+### Handling races between devices.
+
+Even with the tri-state declined flags discussed above, an additional
+complication is that due to networks being unreliable, there's an inherent
+conflict between "what is the current state of enabled/disabled engines?" and
+"what state changes are requested?". For example, if the user changes the state
+of an engine while there's no network, then exits the app, how do we ensure the
+user's new state is updated next time the app starts? What if the user has since
+made a different state request on a different device? Is the state as last-known
+on this device considered canonical?
```

Comment:
I think a three-way merge would be reasonable here, and easy to implement (famous last words 😉)—we're already keeping the last-known engine selections locally anyway; enabling or disabling an engine would flag it as overridden, and we'd merge the changes with other devices on the next sync. I think that would be less surprising than taking the complete set of engine selections, potentially reverting changes on other devices. Especially since engine selections are global, but users might not realize that, and assume they're per-device ("I unchecked history on my work laptop, then bookmarks on my home; why did unchecking bookmarks re-enable history?")

Reply:
I've no real objection to that, but ISTM that either scenario might surprise some users. I agree we should surprise the smallest number of users we can though (but if it seems 50:50, we might as well do the simpler of the things)